### PR TITLE
perf: use indexmap in read::zip_archive::Shared instead of a separate vec and hashmap

### PR DESCRIPTION
**Pull request by Pr0methean**
Replaces #74.

### CODE REVIEW

File: src/read.rs [hunk 1]
```diff
@@ -382,7 +376,7 @@ impl<R: Read + Seek> ZipArchive<R> {
 
         let new_initial_header_start = w.stream_position()?;
         /* Push back file header starts for all entries in the covered files. */
-        new_files.iter_mut().try_for_each(|f| {
+        new_files.iter_mut().try_for_each(|(_name, f)| {
```
Comment (cosmicexplorer) [inline src/read.rs:379]:
```suggestion
        new_files.values_mut().try_for_each(|f| {
```

File: src/write.rs [hunk 2]
```diff
@@ -435,7 +434,7 @@ impl<W: Write + Seek> Write for ZipWriter<W> {
                 if let Ok(count) = write_result {
                     self.stats.update(&buf[0..count]);
                     if self.stats.bytes_written > spec::ZIP64_BYTES_THR
-                        && !self.files.last_mut().unwrap().large_file
+                        && !self.files.last_mut().unwrap().1.large_file
```
Comment (cosmicexplorer) [inline src/write.rs:437]: Could make sense to add a `pub(crate)` accessor like `self.last_file()` to do this sort of thing in a future refactoring PR.
```suggestion
                        && !self.files.last().unwrap().1.large_file
```

**Review by cosmicexplorer — APPROVED** [review]
Thanks so much!!

**Comment by cosmicexplorer** [comment]
Oops, didn't realize this was marked for auto-merge! Please ignore the suggested edits!

**Comment by cosmicexplorer** [comment]
I just rebased and fixed the merge conflicts: you can take my commits at https://github.com/zip-rs/zip2/compare/master...cosmicexplorer:oldpr396a?expand=1 and that should get this passing CI.

**Comment by Pr0methean** [comment]
Replaced by #91.

**Pull request closed without merging**